src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1199 @@
+    // erase it from that set because it's "happy now" (we just processed it).
+    reprocess.erase(tok);
+  }
+
+  size_t max_loop = 1000000, loop_count; // max_loop is to detect epsilon cycles.
+  for (loop_count = 0; !reprocess.empty() && loop_count < max_loop; ++loop_count) {
+    std::vector<Token *> reprocess_vec;
+    for (typename unordered_set<Token *>::iterator iter = reprocess.begin();
+         iter != reprocess.end(); ++iter)
+      reprocess_vec.push_back(*iter);
+    reprocess.clear();
+    for (typename std::vector<Token *>::iterator iter = reprocess_vec.begin();
+         iter != reprocess_vec.end(); ++iter) {
+      Token *tok = *iter;
+      int32 pos = token2pos[tok];
+      // Repeat the processing we did above (for comments, see above).
+      for (ForwardLinkT *link = tok->links; link != NULL; link = link->next) {
+        if (link->ilabel == 0) {
+          IterType following_iter = token2pos.find(link->next_tok);
+          if (following_iter != token2pos.end()) {
+            int32 next_pos = following_iter->second;
+            if (next_pos < pos) {
+              following_iter->second = cur_pos++;
+              reprocess.insert(link->next_tok);
+            }
+          }
+        }
+      }
+    }
+  }
+  KALDI_ASSERT(loop_count < max_loop &&
+               "Epsilon loops exist in your decoding "
+               "graph (this is not allowed!)");
+
+  topsorted_list->clear();
+  topsorted_list->resize(cur_pos, NULL); // create a list with NULLs in between.
+  for (IterType iter = token2pos.begin(); iter != token2pos.end(); ++iter)
+    (*topsorted_list)[iter->second] = iter->first;
+}
+
+template <typename FST, typename Token>
+bool LatticeIncrementalDecoderTpl<FST, Token>::GetLattice(bool use_final_probs,
+                                                          bool redeterminize,
+                                                          CompactLattice *olat) {
+  using namespace fst;
+
+  if (last_get_lattice_frame_ < NumFramesDecoded()) {
+    Lattice raw_fst;
+    // step 1: Get lattice chunk with initial state
+    KALDI_ASSERT(GetRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_,
+                               NumFramesDecoded(), last_get_lattice_frame_ != 0,
+                               !decoding_finalized_));
+    // step 2: Determinize the chunk
+    CompactLattice clat;
+    if (!DeterminizeLatticePhonePrunedWrapper(
+            trans_model_, &raw_fst, config_.lattice_beam, &clat, config_.det_opts))
+      KALDI_WARN << "Determinization finished earlier than the beam";
+
+    final_arc_list_.swap(final_arc_list_prev_);
+    final_arc_list_.clear();
+
+    // step 3.1: Appending new chunk to the old one
+    int32 state_offset = olat->NumStates();
+    if (last_get_lattice_frame_ != 0)
+      state_offset--; // since we do not append initial state
+    unordered_map<int32, size_t>
+        initial_arc_map; // the incoming states of these arcs are initial states
```

Comment:
document what this maps from and to.  There needs to be *much* more documentation here.

Reply:
and your comment is not clear.  "the incoming states of these arcs are initial states".  Initial states in what?  Initial states under what definition?  In an FST?  Or in some other sense?